### PR TITLE
MINIFICPP-449 Add cURL external project build with static linking

### PR DESCRIPTION
**Pull request by achristianson**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -117,6 +118,49 @@ else ()
   message( FATAL_ERROR "OpenSSL was not found. Please install OpenSSL" )
 endif (OPENSSL_FOUND)
 
+if(NOT USE_SYSTEM_CURL)
```
Comment (phrocker) [inline CMakeLists.txt]: This should be in http-curl. If curl is disabled there is no point in adding it as an external project

File: main/CMakeLists.txt [hunk 2]
```diff
@@ -66,7 +66,7 @@ else ()
 	target_link_libraries (minifiexe -Wl,--whole-archive minifi -Wl,--no-whole-archive)
 endif ()
 
-target_link_libraries(minifiexe yaml-cpp ${UUID_LIBRARIES} ${OPENSSL_LIBRARIES})
+target_link_libraries(minifiexe yaml-cpp ${UUID_LIBRARIES} ${CURL_LIBRARIES} ${OPENSSL_LIBRARIES})
```
Comment (phrocker) [inline main/CMakeLists.txt]: This should not be included here. It should be passed transitively through the minifi-http-curl iff it is enabled.

File: extensions/http-curl/CMakeLists.txt [hunk 3]
```diff
@@ -42,9 +42,9 @@ if(CMAKE_THREAD_LIBS_INIT)
 endif()
 
 if (CURL_FOUND)
-        include_directories(${CURL_INCLUDE_DIRS})
-        target_link_libraries (minifi-http-curl ${CURL_LIBRARIES})
-endif(CURL_FOUND)
+  include_directories(${CURL_INCLUDE_DIRS})
+  target_link_libraries(minifi-http-curl ${CURL_LIBRARIES})
```
Comment (phrocker) [inline extensions/http-curl/CMakeLists.txt]: ^ Even in the case of a system library we could rely on the static library if it exists. We don't do it now to ensure we have a small binary size, but we should make it clear that while downloading the source and building manually is an option in most cases a static curl lib likely already exists on the build system. In my case I've just checked my centos 7,6, u16-18, fedora, and suse VMs and they all have static libs. There should be a little work here to ensure we get that IF the user wants that. Perhaps that necessitates an added option?

Comment (achristianson) [reply extensions/http-curl/CMakeLists.txt]: Agreed that we should have the option to static link when not using bundled. I think that should be done in a new ticket that covers all dynamic libs that have static linking options when linking system libs, possibly with a general option e.g. STATIC_LINK_ALL.

**Comment by achristianson** [comment]
Moved to curl extension and fixed external static lib cmake dependencies by using PROPERTIES INTERFACE_LINK_LIBRARIES.

**Pull request closed without merging**